### DISCUSSION TITLE
patch and resume for version 2

### DISCUSSION THREAD

**Discussion by PeterHuang-ST** [elsa-workflows/elsa-core#3202] [General]
I found this answer about patch and resume from this question asked 2 years ago. This is for ELSA V1. Is the work around still the same or does V2 have a different way of patch and resume? Thanks

**Comment by sfmskywalker** [comment]
Hi @PeterHuang-ST , do you have a link to the answer you're referring to for some more context? I'm not entirely sure what is meant with "patch" in this context.

In any case, resuming workflows is fully supported in Elsa 2.

**Reply by PeterHuang-ST** [reply]
Hi Sipke Schoorstra

The link I refer to is this one

https://github.com/elsa-workflows/elsa-core/discussions/358

Patch as if we encountered an error in our code.  Thanks

Peter

**Reply by PeterHuang-ST** [reply]
What would be the ways to performing resuming of the workflow? Would it be performed on a workflow instance or workflow registry display. When we write our code using c#, we would be able to see this in workflow Registry. For V2, would we be able to observe the operation and resume the workflow if the code is hand written in C#? 

For V2, if we design a graphical interface, is there a way of embedded C# code into a activities? If so what would be the way to do it? 

For V2, We would prefer to create complex activities and the workflow graphically and embed C# code for complex activities if possible. I do see an Inline activity under miscellaneous. Can we use this block to run C# code. But if that is not possible, we can write the whole workflow in C#. However, its  more important  if we can observe the workflow graphically if its written in C#. And be able to resume a work flow on error graphically if possible. If that is not possible graphically, please advise on how to perform resuming on a activity? Thanks.

**Reply by sfmskywalker** [reply]
Got it, thanks for the clarification @PeterHuang-ST .

In Elsa 2, you can retry a faulted workflow using the `POST v1/workflow-instances/{workflowInstanceId}/retry` API endpoint, which internally uses the `IWorkflowReviver` service. 

As a user, you can retry faulted workflows from the dashboard app (which uses this endpoint).

![image](https://user-images.githubusercontent.com/938393/179396622-0bb16b9c-4f08-4422-a76f-0390f026623e.png)

> Would it be performed on a workflow instance or workflow registry display

When retrying a faulted workflow, this always applies to a workflow instance.

> When we write our code using c#, we would be able to see this in workflow Registry

Yes, the workflow registry displays workflow definitions from all sources, including C# workflows and workflows created via the designer.

> For V2, would we be able to observe the operation and resume the workflow if the code is hand written in C#?

Yes, all workflow instances are created equal.

> For V2, if we design a graphical interface, is there a way of embedded C# code into a activities? If so what would be the way to do it?

I'm not entirely sure I understand what you mean, but currently the Elsa designer does not support the creation of activities using C# - you will have to do it from your IDE. These activities are then of course available to you from the designer.

> For V2, We would prefer to create complex activities and the workflow graphically and embed C# code for complex activities if possible. I do see an Inline activity under miscellaneous. Can we use this block to run C# code. 

The `Inline` activity should have been hidden from the designer - this is a bug. You cannot create activities from the designer using C# - you can only create custom activities using .cs files using your IDE (or notepad of course) - they need to be compiled as part of your application.
Once compiled, they are available from the dashboard. [This page](https://elsa-workflows.github.io/elsa-core/docs/next/extensibility/extensibility-custom-activities) describes creating custom activities.

> But if that is not possible, we can write the whole workflow in C#

Although you could write the whole workflow in C#, you could also create parts of it in the form of custom activities, and use the designer to connect them together.

> And be able to resume a work flow on error graphically if possible. If that is not possible graphically

As you can see, you can retry faulted workflows from the dashboard.

**Comment by PeterHuang-ST** [comment]
Sipke

Thanks for a thorough answer. I now have much better understanding about ELSA 2.  I will definitely start look into the a custom workflow and see how it can be incorporate into the graphic design. The only question I still have is related to the questions as refer to the link. 

https://github.com/elsa-workflows/elsa-core/discussions/358

specifically, 

    Suppose a workflow has 3 steps. The workflow was faulted at step 2 as program bugs.
    How can I resume this workflow from the step 2 after we fix the program bug?

Basically, if the workflow is blocked because of a bug, we can publish to the work flow with the bug fixes and then resume, correct?

Also, to find out the status of the work flow, we can use the api 

    GET /v1/workflow-instances?workflow=HeartbeatWorkflow&page=0&pageSize=2

to get the list of work flow status. Would a specific workflow instance on the designer shows the current real time running status of a specific long running workflow that was created via the designer? 

Thanks in advance for your answers.

Peter
